akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -142,40 +226,87 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
+    override def handleOutgoingCreated(outStream: OutStream, correlationAttributes: Map[AttributeKey[_], _]): StreamState = {
+      outStream.increaseWindow(extraInitialWindow)
+      Open(buffer, outStream)
     }
-    override def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = SendingData(stream)
-    override def shutdown(): IncomingStreamState = Closed
+    override def handleOutgoingCreatedAndFinished(correlationAttributes: Map[AttributeKey[_], _]): StreamState = HalfClosedLocal(buffer)
+    override def handleOutgoingEnded(): StreamState = Closed
+
+    override protected def onReset(streamId: Int): Unit = multiplexer.closeStream(streamId)
+
+    override def incrementWindow(delta: Int): StreamState = copy(extraInitialWindow = extraInitialWindow + delta)
   }
-  case class ReceivingDataFirst(buffer: IncomingStreamBuffer) extends ReceivingData(HalfClosedRemote) {
-    override def handleOutgoingCreated(stream: Http2SubStream): IncomingStreamState = Open(buffer)
-    override def handleOutgoingEnded(): IncomingStreamState = Closed
+  trait Sending extends StreamState { _: Product =>
+    protected def outStream: OutStream
+
+    override def pullNextFrame(maxSize: Int): (StreamState, PullFrameResult) = {
+      val frame = outStream.nextFrame(maxSize)
+
+      val res =
+        outStream.endStreamIfPossible() match {
+          case Some(trailer) =>
+            PullFrameResult.SendFrameAndTrailer(frame, trailer)
+          case None =>
+            PullFrameResult.SendFrame(frame, outStream.canSend)
+        }
 
-    override protected def onReset(streamId: Int): Unit = multiplexer.cancelSubStream(streamId)
+      val nextState =
+        if (outStream.isDone) handleOutgoingEnded()
+        else this
+
+      (nextState, res)
+    }
+
+    def handleWindowUpdate(windowUpdate: WindowUpdateFrame): StreamState = increaseWindow(windowUpdate.windowSizeIncrement)
+
+    override def handleOutgoingFailed(cause: Throwable): StreamState = Closed
+
+    abstract override def shutdown(): Unit = {
+      outStream.cancelStream()
+      super.shutdown()
+    }
+
+    def increaseWindow(delta: Int): StreamState = {
+      outStream.increaseWindow(delta)
+      this
+    }
   }
-  case class SendingData(outgoingStream: Http2SubStream) extends IncomingStreamState {
-    override def handle(event: StreamFrameEvent): IncomingStreamState = expectIncomingStream(event, HalfClosedRemote, Open, _.withCorrelationAttributes(outgoingStream.correlationAttributes))
-    override def handleOutgoingEnded(): IncomingStreamState = SendingDataHalfClosedLocal(outgoingStream)
-    override def shutdown(): IncomingStreamState = Closed
+
+  case class OpenSendingData(outStream: OutStream, correlationAttributes: Map[AttributeKey[_], _]) extends StreamState with Sending {
+    override def handle(event: StreamFrameEvent): StreamState = event match {
```

Comment:
`handle` doesn't work too well any more because often different traits contribute handling code that needs to be composed. In a future update, we could do the pattern matching in a central place and provide handler methods for each of those, similar to what we already do for `handleWindowUpdate` and `handleResetStream` with ad-hoc solutions.

Reply:
Create low-prio follow up for this issue in https://github.com/akka/akka-http/issues/3576